cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/input/BasicPartitionedFileSetInputContext.java

```diff
@@ -42,34 +49,74 @@
-      throw new IllegalArgumentException(String.format("Expected a '%s', but got '%s'.",
-                                                       FileSplit.class.getName(), inputSplit.getClass().getName()));
+    inputPaths = new ArrayList<>();
+    if (inputSplit instanceof FileSplit) {
+      isCombineInputFormat = false;
+      inputPaths.add(((FileSplit) inputSplit).getPath());
+    } else if (inputSplit instanceof CombineFileSplit) {
+      isCombineInputFormat = true;
+      Collections.addAll(inputPaths, ((CombineFileSplit) inputSplit).getPaths());
+    } else {
+      throw new IllegalArgumentException(String.format("Expected either a '%s' or a '%s', but got '%s'.",
+                                                       FileSplit.class.getName(), CombineFileSplit.class.getName(),
+                                                       inputSplit.getClass().getName()));
     }
-    this.inputPath = ((FileSplit) inputSplit).getPath();
-    this.mappingString = multiInputTaggedSplit.getConf().get(PartitionedFileSetDataset.PATH_TO_PARTITIONING_MAPPING);
+
+    this.conf = multiInputTaggedSplit.getConf();
+    String mappingString = conf.get(PartitionedFileSetDataset.PATH_TO_PARTITIONING_MAPPING);
+    this.pathToPartitionMapping =
+      GSON.fromJson(Objects.requireNonNull(mappingString), STRING_PARTITION_KEY_MAP_TYPE);
   }
 
   @Override
   public PartitionKey getInputPartitionKey() {
-    if (partitionKey == null) {
-      partitionKey = getPartitionKey(inputPath.toUri());
+    if (isCombineInputFormat) {
+      // org.apache.hadoop.mapreduce.lib.input.CombineFileRecordReader sets this in its initNextRecordReader method
+      String inputFileName = conf.get(MRJobConfig.MAP_INPUT_FILE);
+      if (!inputFileName.equals(currentInputfileName)) {
+        Preconditions.checkNotNull(inputFileName);
+        currentPartitionKey = getPartitionKey(URI.create(inputFileName));
+        currentInputfileName = inputFileName;
+      }
+      return currentPartitionKey;
+    } else {
+      List<PartitionKey> inputPartitionKeys = getInputPartitionKeys();
+      if (inputPartitionKeys.size() != 1) {
+        throw new IllegalArgumentException(String.format("Expected only a single PartitionKey, but found: %s",
+                                                         inputPartitionKeys));
+      }
+      return inputPartitionKeys.get(0);
     }
-    return partitionKey;
+  }
+
+  @Override
+  public List<PartitionKey> getInputPartitionKeys() {
+    if (partitionKeys == null) {
+      partitionKeys = new ArrayList<>();
+      for (Path inputPath : inputPaths) {
+        partitionKeys.add(getPartitionKey(inputPath.toUri()));
+      }
+    }
+    return partitionKeys;
   }
 
   private PartitionKey getPartitionKey(URI inputPathURI) {
-    Map<String, PartitionKey> pathToPartitionMapping =
-      GSON.fromJson(Objects.requireNonNull(mappingString), STRING_PARTITION_KEY_MAP_TYPE);
     if (pathToPartitionMapping.containsKey(inputPathURI.toString())) {
```

Comment:
I need to convert it to URI so I can compare only the path component.
I logged the inputPath and the pathToPartitionMapping below to show why its not a simple String prefix check:

```
2017-07-11 20:52:43,793 - INFO  [main:c.c.c.i.a.r.b.d.i.BasicPartitionedFileSetInputContext@92] - alianwar4: inputFileName: hdfs://10.250.0.15/cdap/namespaces/foo/data/rawRecords/files/1499806294856
2017-07-11 20:52:43,793 - INFO  [main:c.c.c.i.a.r.b.d.i.BasicPartitionedFileSetInputContext@93] - alianwar4: mappingString: {"hdfs://xx.xxx.x.xx:8020/cdap/namespaces/foo/data/rawRecords/files/1499806296420":{"time":["java.lang.Long",1499806296420]},"hdfs://xx.xxx.x.xx:8020/cdap/namespaces/foo/data/rawRecords/files/1499806296227":{"time":["java.lang.Long",1499806296227]},"hdfs://xx.xxx.x.xx:8020/cdap/namespaces/foo/data/rawRecords/files/1499806296595":{"time":["java.lang.Long",1499806296595]},"hdfs://xx.xxx.x.xx:8020/cdap/namespaces/foo/data/rawRecords/files/1499806294856":{"time":["java.lang.Long",1499806294856]}}
```